XWIKI-19048: Live data list filter should display and filter by (translated) labels for static list properties

michitux opened the pull request:
This uses the view displayer to populate the metadata that is returned by the REST service.

Jira issue: https://jira.xwiki.org/browse/XWIKI-19048

I would like to get feedback on two questions here:

* The unit test (`StaticListClassPropertyValuesProviderTest`) is not just testing `StaticListClassPropertyValuesProvider` but also its interaction with `ListClass` using an actual `ListClass` object and not a mock. This really helped me during the development to ensure that I'm correctly using the interface and, e.g., handling the XML-encoded result value correctly. However, this is kind of against the idea of an isolated unit test. What do you think, is this kind of test okay or should I change it to, e.g., mock the results of the `ListClass`-methods?
* The implementation of `StaticListClassPropertyValuesProvider::constructPropertyValueForId` is currently quite ugly as it uses the public method `ListClass::displayView`. This involves constructing an additional object with a property just to fulfill the requirements of the parameters and XML-unescaping the result. Its code could be simplified a lot using [`ListClass::getDisplayValue`](https://github.com/xwiki/xwiki-platform/blob/dcfb67ad6b5b78d7d506760ce78d9a3d44b125af/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/ListClass.java#L625), which is currently protected or private, depending on the variant. This method could be used just with the values we already have and a map from values to labels that we can easily obtain and also doesn't require unescaping the result. What do you think about making one of the variants of `ListClass::getDisplayValue` public?

surli commented on xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/classes/StaticListClassPropertyValuesProvider.java:
```diff
@@ -49,13 +51,64 @@
     }
+
+    /**
+     * {@inheritDoc}
```
From what I see there's actually no doc in `AbstractClassPropertyValuesProvider` for it, so maybe it would worth it to add some doc there. Moreover since you override the implem it might also worth it to explain why you override it here.

michitux replied:
Good catch, I've added a comment.

surli commented on xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/classes/StaticListClassPropertyValuesProvider.java:
```diff
@@ -49,13 +51,64 @@
+
+        allValues.stream().map(id -> constructPropertyValueForId(id, propertyDefinition, valueMap))
+            // Filter both by key and by (translated) label (that's why we construct a PropertyValue first)
+            .filter(val -> val.getValue().toString().toLowerCase().contains(lowerFilter)
```
You could have used `StringUtils#containsIgnoreCase` to simplify it.

michitux replied:
Didn't know about that (and it hasn't been used before), I've changed the code to use it.

surli commented:
> What do you think, is this kind of test okay or should I change it to, e.g., mock the results of the ListClass-methods?

So in general better use mock everywhere you can to only test small things. But do ensure that the expectations of what you mock are also tested. So in your case it would mean mocking ListClass and also ensure that the ListClass behaviour you expect are tested. 
Now there's lots of exceptions to this general rule, like if you can't mock some methods because they are static, and we have lots of our code where it's tested like you did.

> What do you think about making one of the variants of ListClass::getDisplayValue public?

+1 to make it public

michitux commented:
Thank you for the review @surli!

I've rebased on master and addressed the comments (in new commits, the existing commits are just rebased). I've also noticed that I missed a `revapi.differences` entry due to the new `@Unstable` annotation, hadn't checked again after adding the annotation. Looking at the code again, I also refactored it a bit to re-use the existing property value from the parent method to avoid that it overrides possibly already existing metadata (no idea if that's possible, in the REST API it's definitely just the query string from the request that is passed as `Object`).

surli commented on xwiki-platform-core/pom.xml:
```diff
@@ -321,6 +321,19 @@
+                  <code>java.annotation.added</code>
+                  <old>method java.lang.String com.xpn.xwiki.objects.classes.ListClass::getDisplayValue(java.lang.String, java.lang.String, java.util.Map&lt;java.lang.String, com.xpn.xwiki.objects.classes.ListItem&gt;, com.xpn.xwiki.XWikiContext) @ com.xpn.xwiki.objects.classes.StaticListClass</old>
+                  <new>method java.lang.String com.xpn.xwiki.objects.classes.ListClass::getDisplayValue(java.lang.String, java.lang.String, java.util.Map&lt;java.lang.String, com.xpn.xwiki.objects.classes.ListItem&gt;, com.xpn.xwiki.XWikiContext) @ com.xpn.xwiki.objects.classes.StaticListClass</new>
+                  <annotation>@org.xwiki.stability.Unstable</annotation>
```
Weird that you only get a revapi problem about the annotation and not about the visibility that has changed. Did you try to build `xwiki-platform-oldcore-legacy` with your changes? It's possible that revapi is only executed there to also include legacy code in the API comparison.

tmortagne replied:
> It's possible that revapi is only executed there

Yes, it's only executed in legacy module.

michitux replied:
It is indeed breaking because I missed all the other child classes of `ListClass` as I didn't try building the legacy module. However, making the method public is not breakage. As discussed in the [chat](https://matrix.to/#/!ikPtGZaGWtyblizzlR:matrix.xwiki.com/$1637144122457QMuoU:matrix.xwiki.com?via=matrix.org&via=matrix.xwiki.com&via=nobis.cloud), I've opened https://github.com/xwiki/xwiki-commons/pull/151 to ignore the `@Unstable`-annotations in general. This makes all of these annotations here obsolete. Should I remove them here or should I still add all the necessary annotations for now such that this can be merged?

surli replied:
> Should I remove them here

you should yes.
